cpp/src/arrow/compute/function.h

```diff
@@ -0,0 +1,203 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// NOTE: API is EXPERIMENTAL and will change without going through a
+// deprecation cycle
+
+#pragma once
+
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "arrow/compute/kernel.h"
+#include "arrow/compute/options.h"  // IWYU pragma: keep
+#include "arrow/result.h"
+#include "arrow/status.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+struct ValueDescr;
+
+namespace compute {
+
+/// \brief Contains the number of required arguments for the function
+struct ARROW_EXPORT FunctionArity {
+  static FunctionArity Nullary() { return FunctionArity(0, false); }
+  static FunctionArity Unary() { return FunctionArity(1, false); }
+  static FunctionArity Binary() { return FunctionArity(2, false); }
+  static FunctionArity Ternary() { return FunctionArity(3, false); }
+  static FunctionArity Varargs(int min_args = 1) { return FunctionArity(min_args, true); }
+
+  FunctionArity(int num_args, bool is_varargs = false)  // NOLINT implicit conversion
```

Comment:
is implicit conversion really necessary here?

Reply:
Well, currently we have code like

```
  auto func = std::make_shared<ScalarFunction>(name, /*arity=*/2);
```

Can't say I'm super fond of writing

```
  auto func = std::make_shared<ScalarFunction>(name, FunctionArity::Binary());
```

but I'm willing to change it

Reply:
I'm renaming FunctionArity -> Arity and will make this refactor now

Reply:
OK that was easy. This is actually fine